### DISCUSSION TITLE
Hybrid property backed by computed column

### DISCUSSION THREAD

**Discussion by couling** [sqlalchemy/sqlalchemy#7241] [Ideas]
I'd really like to see a hybrid between a property and a computed column.  In the example below I use "slug" as a hybrid.  This example shows my current workaround in the absence of a feature.

```python
# This adds "slug" as a computed column and a python generated field and a SQLAlchemy expression
class Foo:
    id: int = sa.Column(sa.Integer, primary_key=True)
    name: str = sa.Column(sa.String(length=128), nullable=False)

    @hybrid_property
    def slug(self) -> str:
        return self.name.replace(' ', '-').lower()

    @slug.expression
    def slug(cls):
        assert cls.__table_args__[0].name == 'slug'
        # the expression is just the fully qualified column name of the computed column
        # This ensures the DB can use the unique index on the column
        return cls.__table_args__[0]

    __table_args__ = (
        # by burying the computed column here, rather than declaring it above, the column will not be queried on SELECT
        # This is true as long as the hybrid_property has the same name
        # So if the expression is costly or the result large, this could be a big performance saving
        sa.Column('slug', sa.Computed("LOWER(REPLACE(name, ' ', '-'))"), unique=True),
    )
```

Very similar to the use case of a hybrid property I want the same property expressed in both SQL and python.  The big difference is that the expression delegates to a computed column *of the same name*.  Amongst other things this lets me add a unique key to the computed column.

-----

The code above is a bit ugly.  But I arrived at it because it covers all aspects of the solution I wanted:
 - sqlalchemy does not add `slug` to the columns of SELECT queries.  In other contexts that can be costly due to computation time or IO
 - `slug` is always calculated in python when directly requested from the python object
 - `slug` may be used in sqlalchemy query expressions and the value will be the SQL generated one
 - `slug` can be indexed.
 - `slug` can be queried by other technologies such as hasura

------

The most dirty thing about the code above (as I see it) is that I can't create a column through declarative mapping without fetching it on every SELECT... that leads me to bury it in __table_args__ which is then a bit dirty to interrogate wen it's needed for the `@hybrid_property.expression`.

-------

I'm not really sure what a feature might look like. I imagine perhaps it would be possible to extend the `hybrid_property` somehow to accept a computed column reference.  

Alternatively it might be possible to extend `Computed` somehow so that it accepts a `lambda` to supply the python generated value.

**Comment by zzzeek** [comment]
this is not hard to do, map the slug column normally as a deferred column, it will never load; key it to an alternate name "_slug", so the hybrid can request it at the expression level only.

what i need to be educated on is why the column needs to be computed in order to place an index on it.   you could have a column called "slug" that's not computed, and just make sure it has that value when you persist it, this can be done in many ways also.   but there's no issue with it being computed either.


```py
class Foo:
    id: int = sa.Column(sa.Integer, primary_key=True)
    name: str = sa.Column(sa.String(length=128), nullable=False)
    _slug: str = deferred(sa.Column('slug', sa.Computed("LOWER(REPLACE(name, ' ', '-'))"), unique=True))

    @hybrid_property
    def slug(self) -> str:
        return self.name.replace(' ', '-').lower()

    @slug.expression
    def slug(cls):
        return cls._slug

```

easy right?

**Reply by couling** [reply]
@zzzeek **Many Thanks** for pointing out `deferred()`, I had missed that even after extensive google searches :-(

But it's right there in the manual so I can't complain.

**Reply by couling** [reply]
@zzzeek 

> what i need to be educated on is why the column needs to be computed in order to place an index

You are right, the justification isn't so clear:  In order to use an index, it must appear exactly in the query.  The important feature of indexing a computed column over indexing an expression is that it trivialises **_sharing_** the precise expression that needs to be included in the query.

> you could have a column called "slug" that's not computed, and just make sure it has that value when you persist it

That assumes the only thing modifying the rows is python.  As noted above that's not the case; we have other things working on the same database. I'd hate to track every usage of this expression outside python just so that I can update every one in every technology when the expression changes.

That leads to options:
 - computed columns
 - triggers - tend to confuse developers and dev-ops alike.
 - functions - don't work well with hasura

**Reply by zzzeek** [reply]
that's fine, if you want the DB to be the source of truth, but if that's a concern then you have to make sure your in -Python rule matches it.    i dont know what kind of scales and real-world schemas / computed structures you are working with but just fetching the column normally is not usually any real overhead.

**Comment by CaselIT** [comment]
Note that depending on the db or computed column configuration, the expression may be run once at insert/update time, so selecting the column is the same as selecting any other column, and that may be faster than re-computing the value python side (and there would be no change of mis-matching computation in python vs db).

I don't think this is something that we should support, since I don't feel it would be a common use case.
Maybe the example provided by Mike above could be added as an example in the library or be placed in the wiki?
